### PR TITLE
Fix #5965: Make scala.quoted.Type poly-kinded

### LINKED ISSUE #5965: Support PCP type healing for higher kinded types
Currently
```scala
def f[T, S[_]](implicit ev: Type[S[T]]): Any = '{
  val a: S[T] = ???
}
```

fails with 
```
15 |      val a: S[T] = ???
   |             ^
   |            access to type S from wrong staging level:
   |             - the definition is at level 0,
   |             - but the access is at level 1.
   |            
   |             The access would be accepted with the right type tag, but
   |             no implicit argument of type scala.quoted.Type[S] was found
15 |      val a: S[T] = ???
   |               ^
   |            access to type T from wrong staging level:
   |             - the definition is at level 0,
   |             - but the access is at level 1.
   |            
   |             The access would be accepted with the right type tag, but
   |             no implicit argument of type scala.quoted.Type[T] was found
```

When healing the types we just do not try to find implicit `Type` for applied higher kinded types. To support this it looks like we need to handle `AppliedTypes` in [PCPCheckAndHeal.checkType](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/Staging.scala#L231) before trying to heal it's parts.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/test/dotty/tools/vulpix/TestConfiguration.scala [hunk 1]
```diff
@@ -47,7 +47,7 @@ object TestConfiguration {
   val yCheckOptions = Array("-Ycheck:all")
 
   val commonOptions = checkOptions ++ noCheckOptions ++ yCheckOptions
-  val defaultOptions = TestFlags(basicClasspath, commonOptions)
+  val defaultOptions = TestFlags(basicClasspath, commonOptions) and "-Ykind-polymorphism" // FIXME remove "-Ykind-polymorphism"
```
Comment (nicolasstucki) [inline compiler/test/dotty/tools/vulpix/TestConfiguration.scala]: The main blocker for this PR is that any code that uses `scala.quote.Type` requires `-Ykind-polymorphism`

**Comment by nicolasstucki** [comment]
To make this work we also need to bootstrap `scala.tasty._`. #5990 is the first step towards this goal.

**Comment by milessabin** [comment]
Very interested to see where this goes.

File: compiler/test/dotty/tools/vulpix/TestConfiguration.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/test/dotty/tools/vulpix/TestConfiguration.scala]: As discussed during today's meeting kind polymorphism is mature enough to be enabled by default. Though I decided to keep a `-Yno-kind-polymorphism` to disable it if needed.

**Comment by nicolasstucki** [comment]
@liufengyun maybe we should wait until #5990 is merged to merge this one. That PR will simplify the changes that are currently in `QuotedOpsImpl.scala` and `QuotedOps.scala`

**Comment by nicolasstucki** [comment]
Also, rebasing #5990 on this one would be more complex than the other way around.

File: library/src-bootstrapped/scala/quoted/Type.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package scala.quoted
+
+import scala.quoted.Types.TaggedType
+import scala.reflect.ClassTag
+import scala.runtime.quoted.Unpickler.Pickled
+
+sealed abstract class Type[T <: AnyKind] {
```
Comment (nicolasstucki) [inline library/src-bootstrapped/scala/quoted/Type.scala]: Here I only added ` <: AnyKind` and `src-non-bootstrapped/scala/quoted/Type.scala` contatins the current version without the bound.

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala [hunk 3]
```diff
@@ -352,8 +352,14 @@ class ReifyQuotes extends MacroTransform {
             transformSplice(spliceTree)
 
           case tree: TypeTree if tree.tpe.typeSymbol.isSplice =>
-            val splicedType = tree.tpe.stripTypeVar.asInstanceOf[TypeRef].prefix.termSymbol
-            transformSplice(ref(splicedType).select(tpnme.splice).withSpan(tree.span))
+            tree.tpe.stripTypeVar match {
+              case tp: AppliedType =>
+                val splicedType = tp.tycon.asInstanceOf[TypeRef].prefix.termSymbol
+                AppliedTypeTree(transformSplice(ref(splicedType).select(tpnme.splice)), tp.args.map(TypeTree(_))).withSpan(tree.span)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: It is not obvious why there is no need to `transformSplice` on the type arguments.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Maybe I should call `transform` on the arguments. I will double check.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Actually, this whole branch is useless. This is already handled by the transformation done in the `Staging` phase.

File: tests/run-with-compiler/i5965.check [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  val y: collection.immutable.List[scala.Int] = scala.List.apply[scala.Int](1, 2, 3)
+  (y: collection.immutable.List[evidence$1$_~$1])
```
Comment (nicolasstucki) [inline tests/run-with-compiler/i5965.check]: `evidence$1$_~$1` was leaking here

**Pull request merged**